File: Universidad de Granada/Programación y Diseño Orientada a Objetos/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#authors: Miguel Ángel Benítez Alguacil
#         Najib Saadouli Arco

require_relative "qytetet.rb"

module ModeloQytetet
  class Jugador
    def initialize(nombre, saldo, encarcelado, cartaLibertad, propiedades, casillaActual, haTiradoDado)
      @nombre = nombre
      @saldo = saldo
      @encarcelado = encarcelado
      @cartaLibertad = cartaLibertad
      @propiedades = propiedades
      @casillaActual = casillaActual
      @haTiradoDado = haTiradoDado
    end
    
    def self.nuevo(nombre)
      self.new(nombre, 7500, false, nil, Array.new, nil, false)
    end
    
    def self.copia(otroJugador)
      self.new(otroJugador.nombre, otroJugador.saldo, otroJugador.encarcelado, otroJugador.cartaLibertad, otroJugador.propiedades, otroJugador.casillaActual, otroJugador.haTiradoDado)
    end
    
    attr_reader :nombre, :saldo, :propiedades
    attr_accessor :encarcelado, :cartaLibertad, :casillaActual, :haTiradoDado
    
    #Otros metodos   
    def cancelarHipoteca(titulo_)
      puedeSerCancelada = tengoSaldo(titulo_.calcularCosteCancelar())
      if(puedeSerCancelada)
        modificarSaldo(-titulo_.calcularCosteCancelar)
        titulo_.cancelarHipoteca()
      end
      return puedeSerCancelada
    end
    
    
    def comprarTituloPropiedad
      costeCompra = @casillaActual.coste
      comprado = false
      if(costeCompra < @saldo)
        @casillaActual.asignarPropietario(self)
        @propiedades << @casillaActual.titulo
        comprado = true
        modificarSaldo(-costeCompra)
      end
      return comprado
    end
    
    def cuantasCasasHotelesTengo
      numCasasHoteles = 0;
      @propiedades.each do |propiedad|
        numCasasHoteles = numCasasHoteles + propiedad.numCasas + propiedad.numHoteles
      end
      return numCasasHoteles
    end
    
    def deboPagarAlquiler
      deboPagar = false
      esMia = esDeMiPropiedad(@casillaActual.titulo)
      if(!esMia)
        tienePropietario = @casillaActual.tengoPropietario()
        if(tienePropietario)
          estaEncarcelado = @casillaActual.propietarioEncarcelado
          if(!estaEncarcelado)
            estaHipotecada = @casillaActual.titulo.hipotecada
            if(!estaHipotecada)
              deboPagar = true
            end
          end
        end
      end
      return deboPagar
    end
    
    def devolverCartaLibertad
        cartaAuxiliar = @cartaLibertad
        @cartaLibertad = nil
        return cartaAuxiliar
    end
    
    def edificarCasa(titulo_)
      edificada = false
      costeEdificarCasa = titulo_.precioEdificar
      
      if(puedoEdificarCasa(titulo_))
        titulo_.edificarCasa
        modificarSaldo(-costeEdificarCasa)
        edificada = true
      end
      
      edificada
    end
    
    def edificarHotel(titulo_)
      edificado = false
      costeEdificarHotel = titulo_.precioEdificar
      
      if(puedoEdificarHotel(titulo_))
        titulo_.edificarHotel()
        self.modificarSaldo(-costeEdificarHotel)
        edificado = true
      end
      
      edificado
    end
    
    def eliminarDeMisPropiedades(titulo_)
      @propiedades.delete(titulo_)
      titulo_.propietario = nil
    end
    
    def esDeMiPropiedad(titulo_)
        esMia = false
        @propiedades.each do |propiedad|
          if(propiedad == titulo_)
            esMia = true
          end
        end
        return esMia
    end
    
    def estoyEnCalleLibre
        raise NotImplementedError
    end
    
    def hipotecarPropiedad(titulo_)
      costeHipoteca = titulo_.hipotecar
      modificarSaldo(costeHipoteca)
    end
    
    def irACarcel(casilla_)
        @casillaActual = casilla_
        @encarcelado = true
    end
    
    def modificarSaldo(cantidad_)
        @saldo = @saldo + cantidad_
        return @saldo
    end
    
    def obtenerCapital()
        valorPropiedad = 0
        capitalTotal = 0
        if(@propiedades != nil)
          @propiedades.each do |propiedad|
            valorPropiedad = valorPropiedad + propiedad.precioCompra + ((propiedad.numCasas + propiedad.numHoteles) * propiedad.precioEdificar)
            if(propiedad.hipotecada)
              valorPropiedad = valorPropiedad - propiedad.hipotecaBase
            end
          end
        end
        capitalTotal = valorPropiedad + @saldo
        return capitalTotal
    end
    
    def obtenerPropiedades(hipotecada_)
        obtenerPropiedades = Array.new
        @propiedades.each do |propiedad|
          if(propiedad.hipotecada == hipotecada_)
            obtenerPropiedades << propiedad
          end
        end
        return obtenerPropiedades
    end
    
    def pagarAlquiler
      costeAlquiler = @casillaActual.pagarAlquiler
      modificarSaldo(-costeAlquiler)
    end
    
    def pagarImpuesto
        modificarSaldo(-@casillaActual.coste)
    end
    
    def pagarLibertad(cantidad_)
        tengoSaldo = tengoSaldo(cantidad_)
        if(tengoSaldo)
          @encarcelado = false
          modificarSaldo(-cantidad_)
        end
    end
    
    def tengoCartaLibertad
        tengoCarta = false
        if(@cartaLibertad != nil)
          tengoCarta = true
        end
        return tengoCarta
    end
    
    def tengoSaldo(cantidad_)
        tengoSaldo = false
        if(@saldo > cantidad_)
          tengoSaldo = true
        end
        return tengoSaldo
    end
    
    def venderPropiedad(casilla_)
        titulo = casilla_.titulo
        eliminarDeMisPropiedades(titulo)
        precioVenta = titulo.calcularPrecioVenta
        modificarSaldo(precioVenta)
    end
    
    def convertirme(fianza)
      Especulador.copia(self, fianza)
    end
    
    def deboIrACarcel
      !tengoCartaLibertad
    end
    
    def puedoEdificarCasa(titulo)
      tengoSaldo(titulo.precioEdificar) && titulo.numCasas < 4
    end
    
    def puedoEdificarHotel(titulo)
      tengoSaldo(titulo.precioEdificar) && titulo.numCasas == 4 && titulo.numHoteles != 4
    end
    
    private :eliminarDeMisPropiedades, :esDeMiPropiedad
    protected :tengoSaldo, :puedoEdificarCasa, :puedoEdificarHotel
    
    def <=>(otroJugador)
      otroJugador.obtenerCapital <=> obtenerCapital
    end
    
    def to_s
      retorno = "\n\tJugador {nombre = #{@nombre}, saldo = #{@saldo}, encarcelado = #{@encarcelado}, cartaLibertad = #{@cartaLibertad}, casillaActual = #{@casillaActual}, propiedades = "
      @propiedades.each do |propiedad|    # saldo = #{obtenerCapital},
        retorno << propiedad.to_s 
      end
      retorno << ",\n\tcapital del jugador = #{obtenerCapital}}\n"
    end
    
  end
end

File: Universidad de Granada/Programación y Diseño Orientada a Objetos/QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;

/**
 *
 * @author Miguel Ángel Benítez Alguacil
 *         Najib Saadouli Arco
 */
public class Jugador implements java.lang.Comparable {
    private boolean encarcelado;
    private String nombre;
    private int saldo;
    private Sorpresa cartaLibertad;
    private ArrayList<TituloPropiedad> propiedades;
    private Casilla casillaActual;
    private boolean haTiradoDado;

    public Jugador(String nombre) {
        this.nombre = nombre;
        propiedades = new ArrayList<>();
        casillaActual = null;
        haTiradoDado = false;
        cartaLibertad = null;
        saldo = 7500;
        encarcelado = false;
    }
    
    protected Jugador(Jugador otroJugador){
        nombre = otroJugador.nombre;
        propiedades = otroJugador.propiedades;
        casillaActual = otroJugador.casillaActual;
        haTiradoDado = otroJugador.haTiradoDado;
        cartaLibertad = otroJugador.cartaLibertad;
        saldo = otroJugador.saldo;
        encarcelado = otroJugador.encarcelado;
    }
    
    //Getters de los atributos de Jugador
    boolean getEncarcelado() {
        return encarcelado;
    }

    public String getNombre() {
        return nombre;
    }

    public int getSaldo() {
        return saldo;
    }

    Sorpresa getCartaLibertad() {
        return cartaLibertad;
    }

    ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    Casilla getCasillaActual() {
        return casillaActual;
    }

    //Setters de Jugador
    void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }

    void setSaldo(int saldo) {
        this.saldo = saldo;
    }

    void setCartaLibertad(Sorpresa cartaLibertad) {
        this.cartaLibertad = cartaLibertad;
    }

    void setCasillaActual(Casilla casillaActual) {
        this.casillaActual = casillaActual;
    }
    
    public void setHatiradoDado(boolean haTiradoDado){
        this.haTiradoDado = haTiradoDado;
    }
    
    //Otros metodos
    boolean cancelarHipoteca(TituloPropiedad titulo){
        boolean puedeSerCancelada = tengoSaldo(titulo.calcularCosteCancelar());
        if(puedeSerCancelada)
            modificarSaldo(-titulo.calcularCosteCancelar());
            titulo.cancelarHipoteca();
        return puedeSerCancelada;
    }
    
    boolean comprarTituloPropiedad(){
        int costeCompra = getCasillaActual().getCoste();
        boolean comprado = false;
        if (costeCompra < getSaldo()){
            ((Calle)casillaActual).asignarPropietario(this);
            propiedades.add(getCasillaActual().getTitulo());
            comprado = true;
            modificarSaldo(-costeCompra);
        }
        return comprado;
    }
    
    int cuantasCasasHotelesTengo(){
        int numeroTotalCasasHoteles = 0;
        for (int i = 0; i < propiedades.size(); i++){
            numeroTotalCasasHoteles += propiedades.get(i).getNumCasas();
            numeroTotalCasasHoteles += propiedades.get(i).getNumHoteles();        
        }
        return numeroTotalCasasHoteles;
    }
    
    boolean deboPagarAlquiler(){
        TituloPropiedad titulo = casillaActual.getTitulo();
        boolean depoPagar = false, tienePropietario, estaEncarcelado, estaHipotecada, esDeMiPropiedad = esDeMiPropiedad(titulo);
        if(!esDeMiPropiedad){
            tienePropietario = ((Calle)casillaActual).tengoPropietario();
            if(tienePropietario){
                estaEncarcelado = ((Calle)casillaActual).propietarioEncarcelado();
                if(!estaEncarcelado){
                    estaHipotecada = titulo.getHipotecada();
                    if(!estaHipotecada)
                        depoPagar = true;
                }
            }
        }
        return depoPagar;
    }
    
    Sorpresa devolverCartaLibertad(){
        Sorpresa cartaAuxiliar = getCartaLibertad();
        setCartaLibertad(null);        
        return cartaAuxiliar;
    }
    
    boolean edificarCasa(TituloPropiedad titulo){
        boolean edificada = false;
        int costeEdificarCasa = titulo.getPrecioEdificar();
        
        if(this.puedoEdificarCasa(titulo)){
            titulo.edificarCasa();
            modificarSaldo(-costeEdificarCasa);
            edificada = true;
        }
        
        return edificada;
    }
    
    boolean edificarHotel(TituloPropiedad titulo){
        boolean edificado = false;
        int costeEdificarHotel = titulo.getPrecioEdificar();
        
        if(this.puedoEdificarHotel(titulo)){
            titulo.edificarHotel();
            modificarSaldo(-costeEdificarHotel);
            edificado = true;
        }
        
        return edificado;
    }
    
    private void eliminarDeMisPropiedades(TituloPropiedad titulo){
        propiedades.remove(titulo);
        titulo.setPropietario(null);
    }
    
    private boolean esDeMiPropiedad(TituloPropiedad titulo){
        boolean deMiPropiedad = false;
        for(TituloPropiedad propiedad: propiedades){
            if(titulo == propiedad){
                deMiPropiedad = true;
            }
        }
        return deMiPropiedad;
    }
    
    boolean estoyEnCalleLibre(){
        throw new UnsupportedOperationException("no implementado");
    }
    
    void hipotecarPropiedad(TituloPropiedad titulo){
        int costeHipoteca = titulo.hipotecar();
        modificarSaldo(costeHipoteca);
    }
    
    void irACarcel(Casilla casilla){
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    void modificarSaldo(int cantidad){
        saldo += cantidad;
    }
    
    int obtenerCapital(){
        int valorPropiedad = 0, capitalTotal = 0;
        if(propiedades!=null){
            for(int i=0; i < propiedades.size(); i++){
                valorPropiedad += (propiedades.get(i).getNumCasas() + propiedades.get(i).getNumHoteles()) * propiedades.get(i).getPrecioEdificar();
                if (propiedades.get(i).getHipotecada())
                    valorPropiedad -= propiedades.get(i).getHipotecaBase();
            }
        }
        capitalTotal = valorPropiedad + getSaldo();
        return capitalTotal;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean hipotecada){
        ArrayList<TituloPropiedad> propiedadesHipotecada = new ArrayList<>();
        for(int i=0; i<getPropiedades().size(); i++)
            if(getPropiedades().get(i).getHipotecada() == hipotecada)
                propiedadesHipotecada.add(getPropiedades().get(i));
        return propiedadesHipotecada;
    }
    
    void pagarAlquiler(){
        int costeAlquiler = ((Calle)casillaActual).pagarAlquiler();
        modificarSaldo(-costeAlquiler);
    }
    
    protected void pagarImpuesto(){
        modificarSaldo(-getCasillaActual().getCoste());
    }
    
    void pagarLibertad(int cantidad){
        boolean tengoSaldo = tengoSaldo(cantidad);
        if(tengoSaldo){
            setEncarcelado(false);
            modificarSaldo(-cantidad);
        }
    }
    
    boolean tengoCartaLibertad(){
        boolean tengoCarta = false;
        if(getCartaLibertad() != null)
            tengoCarta = true;
        return tengoCarta;
    }
    
    protected boolean tengoSaldo(int cantidad){
        boolean tengo = false;
        if(getSaldo() > cantidad)
            tengo = true;
        return tengo;
    }
    
    void venderPropiedad(Casilla casilla){
        TituloPropiedad titulo = casilla.getTitulo();
        eliminarDeMisPropiedades(titulo);
        int precioVenta = titulo.calcularPrecioVenta();
        modificarSaldo(precioVenta);
    }
    
    protected Especulador convertirme(int fianza){
        Especulador especulador = new Especulador(this, fianza);
        return especulador;
    }
    
    protected boolean deboIrACarcel (){
        return !tengoCartaLibertad();
    }
    
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        return ( tengoSaldo(titulo.getPrecioEdificar()) && titulo.getNumCasas()<4 );
    }
    
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        return ( tengoSaldo(titulo.getPrecioEdificar()) && titulo.getNumCasas()==4 && titulo.getNumHoteles()!=4 );
    }
    
    @Override
    public int compareTo(Object otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return (otroCapital - obtenerCapital());
    }
    
    @Override
    public String toString() {
        return "\n\tJugador{" + "nombre = " + nombre + ", encarcelado = " + 
            encarcelado + ", saldo = " + saldo + ", carta libertad = " + cartaLibertad
                + ", casilla actual = " + casillaActual + ", propiedades = " + propiedades
                + ", capital del jugador = " + obtenerCapital() 
                + "}";
    }
    
}
